src/Storages/NATS/NATSConnection.cpp

```diff
@@ -9,29 +9,66 @@
 namespace DB
 {
 
-static const auto RETRIES_MAX = 20;
-static const auto CONNECTED_TO_BUFFER_SIZE = 256;
+/// disconnectedCallback may be called after connection destroy
+LoggerPtr NATSConnection::callback_logger = getLogger("NATSConnection callback");
 
-
-NATSConnectionManager::NATSConnectionManager(const NATSConfiguration & configuration_, LoggerPtr log_)
+NATSConnection::NATSConnection(const NATSConfiguration & configuration_, LoggerPtr log_, NATSOptionsPtr options_)
     : configuration(configuration_)
-    , log(log_)
-    , event_handler(loop.getLoop(), log)
+    , log(std::move(log_))
+    , options(std::move(options_))
+    , connection(nullptr, &natsConnection_Destroy)
 {
+    if (!configuration.username.empty() && !configuration.password.empty())
+        natsOptions_SetUserInfo(options.get(), configuration.username.c_str(), configuration.password.c_str());
+    if (!configuration.token.empty())
+        natsOptions_SetToken(options.get(), configuration.token.c_str());
+    if (!configuration.credential_file.empty())
+        natsOptions_SetUserCredentialsFromFiles(options.get(), configuration.credential_file.c_str(), nullptr);
+
+    if (configuration.secure)
+    {
+        natsOptions_SetSecure(options.get(), true);
+    }
+
+    // use CLICKHOUSE_NATS_TLS_SECURE=0 env var to skip TLS verification of server cert
     const char * val = std::getenv("CLICKHOUSE_NATS_TLS_SECURE"); // NOLINT(concurrency-mt-unsafe) // this is safe on Linux glibc/Musl, but potentially not safe on other platforms
     std::string tls_secure = val == nullptr ? std::string("1") : std::string(val);
     if (tls_secure == "0")
-        skip_verification = true;
-}
+    {
+        natsOptions_SkipServerVerification(options.get(), true);
+    }
 
+    if (!configuration.url.empty())
+    {
+        natsOptions_SetURL(options.get(), configuration.url.c_str());
+    }
+    else
+    {
+        const char * servers[configuration.servers.size()];
+        for (size_t i = 0; i < configuration.servers.size(); ++i)
+        {
+            servers[i] = configuration.servers[i].c_str();
+        }
+        natsOptions_SetServers(options.get(), servers, static_cast<int>(configuration.servers.size()));
+    }
 
-NATSConnectionManager::~NATSConnectionManager()
+    static constexpr int infinite_reconnect_count = -1;
+    natsOptions_SetMaxReconnect(options.get(), infinite_reconnect_count);
+    natsOptions_SetReconnectWait(options.get(), configuration.reconnect_wait);
+    natsOptions_SetDisconnectedCB(options.get(), disconnectedCallback, this);
+    natsOptions_SetReconnectedCB(options.get(), reconnectedCallback, this);
+}
+NATSConnection::~NATSConnection()
 {
-    if (has_connection)
-        natsConnection_Destroy(connection);
+    std::lock_guard lock(mutex);
+    if (!isDisconnectedImpl(lock))
+    {
+        natsConnection_Close(connection.get());
+    }
```

Comment:
Shouldn't this work?
```suggestion
    disconnect()
```